contracts/child/ChildToken/ChildMintableERC721.sol

```diff
@@ -0,0 +1,92 @@
+    bytes32 public constant DEPOSITOR_ROLE = keccak256("DEPOSITOR_ROLE");
+    mapping (uint256 => bool) public withdrawnTokens;
+
+    constructor(
+        string memory name_,
+        string memory symbol_
+    ) public ERC721(name_, symbol_) NetworkAgnostic(name_, ERC712_VERSION, ROOT_CHAIN_ID) {
+        _setupRole(DEFAULT_ADMIN_ROLE, _msgSender());
+        _setupRole(DEPOSITOR_ROLE, _msgSender());
+    }
+
+    modifier only(bytes32 role) {
+        require(hasRole(role, _msgSender()), "ChildMintableERC721: INSUFFICIENT_PERMISSIONS");
+        _;
+    }
+
+    function _msgSender()
+        internal
+        override
+        view
+        returns (address payable sender)
+    {
+        if (msg.sender == address(this)) {
+            bytes memory array = msg.data;
+            uint256 index = msg.data.length;
+            assembly {
+                // Load the 32 bytes word from memory with the address on the lower 20 bytes, and mask those.
+                sender := and(
+                    mload(add(array, index)),
+                    0xffffffffffffffffffffffffffffffffffffffff
+                )
+            }
+        } else {
+            sender = msg.sender;
+        }
+        return sender;
+    }
+
+    /**
+     * @notice called when token is deposited on root chain
+     * @dev Should be callable only by ChildChainManager
+     * Should handle deposit by minting the required tokenId for user
+     * Should set `withdrawnTokens` mapping to `false` for the tokenId being deposited
+     * Minting can also be done by other functions
+     * @param user user address for whom deposit is being done
+     * @param depositData abi encoded tokenId
+     */
+    function deposit(address user, bytes calldata depositData)
+        external
+        override
+        only(DEPOSITOR_ROLE)
+    {
+        uint256 tokenId = abi.decode(depositData, (uint256));
+        withdrawnTokens[tokenId] = false;
+        _mint(user, tokenId);
+    }
+
+    /**
+     * @notice called when user wants to withdraw token back to root chain
+     * @dev Should handle withraw by burning user's token.
+     * Should set `withdrawnTokens` mapping to `true` for the tokenId being withdrawn
+     * This transaction will be verified when exiting on root chain
+     * @param tokenId tokenId to withdraw
+     */
+    function withdraw(uint256 tokenId) external {
+        require(_msgSender() == ownerOf(tokenId), "ChildMintableERC721: INVALID_TOKEN_OWNER");
+        withdrawnTokens[tokenId] = true;
+        _burn(tokenId);
+    }
+
+    /**
+     * @notice Example function to handle minting tokens on matic chain
+     * @dev Minting can be done as per requirement
+     * Should verify if token is withdrawn by checking `withdrawnTokens` mapping
+     * @param user user for whom tokens are being minted
+     * @param tokenId tokenId to mint
+     */
+    function mint(address user, uint256 tokenId) public {
```

Comment:
Should we add `only(role)` just to make sure people don't directly deploy it?

Reply:
Adding `only(DEFAULT_ADMIN_ROLE)` and mentioning in comment.